lib/ryu.fz

```diff
@@ -0,0 +1,994 @@
+      is
+        exponent*78913 >> 18
+
+
+      q := max (i64 0) (floor_log₁₀_pow_2 exponent)-1
+      k := pow5_inv_bitcount + pow5bits q
+      i := -exponent + q + k
+
+      dm := mul_pow5_inv_div_pow2 smaller_halfway_point q i
+      dv := mul_pow5_inv_div_pow2 mv q i
+      dp := mul_pow5_inv_div_pow2 larger_halfway_point q i
+
+      if q <= 21
+        if mv % 5 = 0
+          (dv, dp, dm, false, (multiple_of_power_of_5 mv q), q)
+        else if rounding_mode.accept_upper_bound is_even
+          (dv, dp, dm, (multiple_of_power_of_5 smaller_halfway_point q), false,  q)
+        else if multiple_of_power_of_5 larger_halfway_point q
+          (dv, dp-1, dm, false, false, q)
+        else
+          (dv, dp, dm, false, false, q)
+      else
+        (dv, dp, dm, false, false, q)
+    else
+
+
+      # computes ⌊log₁₀ 5ᵉ⌋
+      floor_log₁₀_pow_5 (e i64) i64
+      pre e >= 0,
+          e <= 2620
+      post
+        ( f64s.log 10 (5.0 ** e.as_f64) = f64s.infinity
+          || floor_log₁₀_pow_5.this.result = (f64s.log 10 (5.0 ** e.as_f64)).floor.as_i64)
+      is
+        e*732923 >> 20
+
+
+      q := max (i64 0) (floor_log₁₀_pow_5 -exponent)-1
+      i := -exponent - q
+      k := pow5bits i + 1 - pow5_bitcount
+      j := q - k
+
+      dm := mul_pow5_div_pow2 smaller_halfway_point i j
+      dv := mul_pow5_div_pow2 mv i j
+      dp := mul_pow5_div_pow2 larger_halfway_point i j
+      e10 := q + exponent
+
+      if q <= 1
+        if rounding_mode.accept_upper_bound is_even
+          (dv, dp, dm, mm_shift, true, e10)
+        else
+          (dv, dp -  1, dm, false, true, e10)
+      else if q < 63
+        (dv, dp, dm, false, (mv & (((i64 1) << (q - 1)) - 1)) = 0, e10)
+      else
+        (dv, dp, dm, false, false, e10)
+
+
+
+  # step 4: shortest decimal representation
+  private step_4(dv, dp, dm i64, dm_is_trailing_zeros, dv_is_trailing_zeros bool, e10 i64, even, scientific_notation bool, rounding_mode rounding_mode) tuple i64 i64 i64 is
+
+
+    # get the decimal length of this i64
+    private decimal_length(v i64) i64 is
+      if      v >= 1000000000000000000 then 19
+      else if v >= 100000000000000000  then 18
+      else if v >= 10000000000000000   then 17
+      else if v >= 1000000000000000    then 16
+      else if v >= 100000000000000     then 15
+      else if v >= 10000000000000      then 14
+      else if v >= 1000000000000       then 13
+      else if v >= 100000000000        then 12
+      else if v >= 10000000000         then 11
+      else if v >= 1000000000          then 10
+      else if v >= 100000000           then  9
+      else if v >= 10000000            then  8
+      else if v >= 1000000             then  7
+      else if v >= 100000              then  6
+      else if v >= 10000               then  5
+      else if v >= 1000                then  4
+      else if v >= 100                 then  3
+      else if v >= 10                  then  2
+      else                                   1
```

Comment:
Might help to do this the other way around, i.e., start checking `v < 10`, then `v < 100`, etc. to shorten the most common paths.